### PR TITLE
Allow storm functions to call each other.

### PR DESCRIPTION
**Pull request by rakuy0**

### CODE REVIEW

File: synapse/tests/test_lib_ast.py [hunk 1]
```diff
@@ -651,6 +651,110 @@ class AstTest(s_test.SynTest):
...
+            $lib.print("got {out}", out=$output)
+            '''
+            msgs = await core.streamstorm(q).list()
+            expected = ['this should be first', 'toreturn called', 'got foobar']
+            prints = list(filter(lambda m: m[0] == 'print', msgs))
+            self.eq(len(prints), len(expected))
+            for idx, out in enumerate(expected):
+                self.stormIsInPrint(out, msgs)
+                self.eq(out, prints[idx][1]['mesg'])
+            self.stormIsInPrint('this should be first', msgs)
```
Comment (Cisphyx) [inline synapse/tests/test_lib_ast.py]: Seems like some redundant tests got left here.

Comment (rakuy0) [reply synapse/tests/test_lib_ast.py]: :+1:

File: synapse/tests/test_lib_ast.py [hunk 2]
```diff
@@ -651,6 +700,204 @@ class AstTest(s_test.SynTest):
...
+            q = '''
+            $count = 0
+            function baz(arg3) {
+                [ test:str = $arg3 ]
+                $count = $( $count + 1)
+                [ test:str = "cool" ]
+            }
+
+            function bar(arg2) {
+                yield $baz($arg2)
```
Comment (vEpiphyte) [inline synapse/tests/test_lib_ast.py]: I thought we gutted the yield keyword from the function syntax?

Comment (rakuy0) [reply synapse/tests/test_lib_ast.py]: Maybe it's still in PR? Because this syntax still yields out the node I expect.

File: synapse/lib/storm.py [hunk 3]
```diff
@@ -390,11 +396,50 @@ def loadRuntVars(self, query):
                 self.tick()
                 yield node, path
 
-    async def getScopeRuntime(self, query, opts=None):
+    async def getScopeRuntime(self, query, opts=None, impd=False, modl=False):
```
Comment (vEpiphyte) [inline synapse/lib/storm.py]: This needs docstrings

File: synapse/lib/storm.py [hunk 4]
```diff
@@ -390,11 +396,50 @@ def loadRuntVars(self, query):
...
+            runt.modulefuncs = dict(self.modulefuncs)
+            for name, oper in self.modulefuncs.items():
+                runt.runtvars.add(name)
+                async for item in oper.run(runt, s_ast.agen()):
+                    pass  # pragma: no cover
+
         runt.loadRuntVars(query)
         return runt
 
+    async def propBackGlobals(self, runt):
```
Comment (vEpiphyte) [inline synapse/lib/storm.py]: needs docstring - we haven't used `prop` as a function prefix before to indicate that we need to do data propagation

File: synapse/lib/storm.py [hunk 5]
```diff
@@ -390,11 +396,50 @@ def loadRuntVars(self, query):
...
+        for name in runt.globals:
+            if name in self.modulefuncs or name == 'lib':
+                # don't override our parent's version of a function
+                continue
+            self.vars[name] = runt.vars[name]
```
Comment (vEpiphyte) [inline synapse/lib/storm.py]: This implies that the keys of runt.vars is a subset of runt.globals.  That **Should** be the case but a `.get(name, s_common.novalu)`, check and set can avoid a keyerror.

File: synapse/lib/storm.py [hunk 6]
```diff
@@ -234,6 +234,10 @@ def __init__(self, snap, opts=None, user=None):
         self.runtvars = set()
         self.runtvars.update(self.vars.keys())
         self.runtvars.update(self.ctors.keys())
+        self.isModuleRunt = False
+        self.isImport = False
```
Comment (vEpiphyte) [inline synapse/lib/storm.py]: `self.isImport` appears to be a unused class variable. is this a vestigial implementation element?

File: synapse/lib/storm.py [hunk 7]
```diff
@@ -390,11 +396,50 @@ def loadRuntVars(self, query):
                 self.tick()
                 yield node, path
 
-    async def getScopeRuntime(self, query, opts=None):
+    async def getScopeRuntime(self, query, opts=None, impd=False, modl=False):
         runt = Runtime(self.snap, user=self.user, opts=opts)
+        runt.isModuleRunt = modl
+        runt.isImport = impd
+        if not impd:  # respect the import boundary
```
Comment (vEpiphyte) [inline synapse/lib/storm.py]: is this where runt.isImport should be used?

File: synapse/lib/snap.py [hunk 8]
```diff
@@ -88,6 +88,7 @@ def getStormRuntime(self, opts=None, user=None):
             user = self.user
 
         runt = s_storm.Runtime(self, opts=opts, user=user)
+        runt.isModuleRunt = True
```
Comment (vEpiphyte) [inline synapse/lib/snap.py]: so every runtime made directly from the snap itself is a modulerunt. is there ever a case where we wouldn't want that?

File: synapse/lib/storm.py (hunk 3 above)
Comment (rakuy0) [reply synapse/lib/storm.py]: :+1:

File: synapse/lib/storm.py (hunk 4 above)
Comment (rakuy0) [reply synapse/lib/storm.py]: Okay. Is there a more conventional verb that would be in line?

File: synapse/lib/storm.py (hunk 6 above)
Comment (rakuy0) [reply synapse/lib/storm.py]: Ah, yes. it was from one of the initial attempts. I'll remove it.

File: synapse/lib/storm.py (hunk 7 above)
Comment (rakuy0) [reply synapse/lib/storm.py]: At one point yes, and also down in propBackGlobals, but that also turned out to be unneeded. I will clean up now.

File: synapse/lib/storm.py (hunk 5 above)
Comment (rakuy0) [reply synapse/lib/storm.py]: :+1:

File: synapse/lib/snap.py (hunk 8 above)
Comment (rakuy0) [reply synapse/lib/snap.py]: Hmmm. If there's no function calls or definitions, isModuleRunt has no effect. If there's only function definitions, those will get recorded but nothing beyond that. So outside of functions, this won't really impact anything. And since runts are specfic to a particular query instance, there's no pollution of isModuleRunt across runtimes.

TL;DR: I can't think of one.

File: synapse/tests/test_lib_ast.py [hunk 9]
```diff
@@ -651,6 +696,248 @@ class AstTest(s_test.SynTest):
...
+                return ($toreturn)
+            }
+
+            $func = $wrapper()
+            $lib.print("this should be first")
+            $output = $func()
+            $lib.print("got {out}", out=$output)
+            '''
+            msgs = await core.streamstorm(q).list()
+            expected = ['this should be first', 'toreturn called', 'got foobar']
```
Comment (vEpiphyte) [inline synapse/tests/test_lib_ast.py]: why not just call stormIsInPrint 3 times?

File: synapse/lib/storm.py [hunk 10]
```diff
@@ -390,11 +395,64 @@ def loadRuntVars(self, query):
                 self.tick()
                 yield node, path
 
-    async def getScopeRuntime(self, query, opts=None):
+    async def getScopeRuntime(self, query, opts=None, impd=False, modl=False):
```
Comment (vEpiphyte) [inline synapse/lib/storm.py]: clean up call sig

Comment (rakuy0) [reply synapse/lib/storm.py]: :+1:

File: synapse/tests/test_lib_ast.py (hunk 9 above)
Comment (rakuy0) [reply synapse/tests/test_lib_ast.py]: fixed. it was supposed to testing the order of the messages, and it was, but i made it more robust.

File: synapse/lib/storm.py [hunk 11]
```diff
@@ -395,11 +400,65 @@ def loadRuntVars(self, query):
...
+                    valu = self.vars.get(name, s_common.novalu)
+                    if valu is s_common.novalu:
+                        continue
+                    if name not in self.modulefuncs and name != 'lib':
+                        runt.globals.add(name)
+                        runt.runtvars.add(name)
+                        runt.vars[name] = valu
+
+            # propagate down all the global variables
+            for name in self.globals:
```
Comment (invisig0th) [inline synapse/lib/storm.py]: how is globals different than vars?

Comment (rakuy0) [reply synapse/lib/storm.py]: It's to denote which variables we received from the outer scope, so that we know which ones whose values we can propagate back up and which to avoid any kind of variable pollution.
